File: rust/stringzilla/overlap.rs
//! Window overlap between a batch of prepared queries and any number of candidate batches.

use super::*;
use core::ffi::c_void;
use core::mem::MaybeUninit;

/// A forest of prepared query trees, probed by as many batches of candidates as a caller has.
///
/// A window is a fixed-width byte n-gram, and the overlap of two texts at that width is the count of
/// one's windows that occur in the other, over whichever of the two holds more. Every query is hashed
/// and sorted into one B-tree over every width at construction, so the sort is paid once per batch
/// rather than once per round, and nothing is stored per candidate.
///
/// The score is asymmetric: a candidate's window occurrences count against a query's distinct windows,
/// so swapping the two sides changes the answer whenever either repeats a window.
///
/// The fields mirror `sz_overlap_engine_t` one for one and only `count` and `widths_count` are read
/// from Rust, so the layout is load-bearing and the engine travels to C by pointer. Owning raw
/// pointers makes it neither `Send` nor `Sync`, which is what the C contract wants: a compute verb
/// grows the engine's round scratch, so it mutates, and [`OverlapEngine::scores`] takes `&mut self`
/// for that reason.
///
/// # Examples
///
/// ```rust
/// use stringzilla::sz::OverlapEngine;
///
/// let mut engine = OverlapEngine::new(&["the quick brown fox"], &[4, 8]).unwrap();
///
/// // A `[queries, candidates, widths]` block with the width axis unit-strided.
/// let mut scores = [0.0f32; 2];
/// engine.scores(&["the quick brown cat"], &mut scores, 2, 2).unwrap();
/// assert!(scores.iter().all(|share| (0.0..=1.0).contains(share)));
/// ```
#[repr(C)]
#[allow(dead_code)] // Every member is the C side's to write; the layout is what Rust keeps.
pub struct OverlapEngine {
    nodes: *const u32,
    nodes_offsets: *const usize,
    keys_counts: *const u32,
    widths: *const u32,
    powers: *const u32,
    lengths: *const u32,
    count: usize,
    widths_count: usize,
    capability: i32,
    alloc: _SzMemoryAllocator,
    memory: *mut c_void,
    memory_bytes: usize,
    scratch: *mut c_void,
    scratch_bytes: usize,
}

impl OverlapEngine {
    /// Sorts `queries` into one forest on the host, resolving the CPU tier once.
    ///
    /// `window_widths` are n-gram widths in bytes and need not form a doubling chain; a width past a
    /// text scores zero for every pair it spans. Zero widths are refused.
    pub fn new<Query>(queries: &[Query], window_widths: &[usize]) -> Result<Self, Status>
    where
        Query: AsRef<[u8]>,
    {
        let mut engine = MaybeUninit::<Self>::uninit();
        let status = with_sequence(queries, |sequence| unsafe {
            sz_overlap_engine_init_cpu(
                sequence,
                window_widths.as_ptr(),
                window_widths.len(),
                core::ptr::null(),
                engine.as_mut_ptr(),
            )
        });
        match status {
            Status::Success => Ok(unsafe { engine.assume_init() }),
            error => Err(error),
        }
    }

    /// Sorts `queries` into one forest on `stream`'s device, resolving the launch geometry once.
    ///
    /// `stream` is a `cudaStream_t`, or null for the current device's default one. A width the device
    /// backend's per-thread ring cannot hold is refused here rather than at the first round.
    ///
    /// A compute verb of a device engine also needs a candidate sequence whose accessors run on the
    /// device, which this crate cannot build yet, so such an engine is constructible here before it
    /// is drivable and every verb below answers `Status::DeviceMemoryMismatch` for it.
    ///
    /// # Safety
    ///
    /// `stream` must be a live stream of the current context, and it makes every later verb of this
    /// engine asynchronous: [`OverlapEngine::scores`] enqueues and returns, so its output slice has to
    /// be device-reachable, has to outlive the launch, and must not be read before the caller joins
    /// `stream` itself.
    #[cfg(feature = "cuda")]
    pub unsafe fn new_on_gpu<Query>(
        queries: &[Query],
        window_widths: &[usize],
        stream: *mut c_void,
    ) -> Result<Self, Status>
    where
        Query: AsRef<[u8]>,
    {
        let mut engine = MaybeUninit::<Self>::uninit();
        let status = with_sequence(queries, |sequence| unsafe {
            sz_overlap_engine_init_gpu(
                sequence,
                window_widths.as_ptr(),
                window_widths.len(),
                core::ptr::null(),
                stream,
                engine.as_mut_ptr(),
            )
        });
        match status {
            Status::Success => Ok(unsafe { engine.assume_init() }),
            error => Err(error),
        }
    }

    /// Queries the forest holds, which is the first axis of every output.
    pub fn queries_count(&self) -> usize {
        self.count
    }

    /// Window widths the batch scores at, which is the last axis of every output.
    pub fn widths_count(&self) -> usize {
        self.widths_count
    }

    /// Window overlap of every prepared query with every candidate, at every width of the batch.
    ///
    /// `scores` receives a `[queries, candidates, widths]` block: share `[q, c, w]` lands at
    /// `scores[q * scores_query_stride + c * scores_candidate_stride + w]`, each in `[0, 1]`, with the
    /// width axis unit-strided. Both strides count entries rather than bytes, the candidate stride
    /// being at least the width count and the query stride at least the candidates times that.
    pub fn scores<Candidate>(
        &mut self,
        candidates: &[Candidate],
        scores: &mut [f32],
        scores_query_stride: usize,
        scores_candidate_stride: usize,
    ) -> Result<(), Status>
    where
        Candidate: AsRef<[u8]>,
    {
        if scores_candidate_stride < self.widths_count {
            return Err(Status::UnexpectedDimensions);
        }
        let candidates_span = candidates
            .len()
            .checked_mul(scores_candidate_stride)
            .ok_or(Status::OverflowRisk)?;
        if scores_query_stride < candidates_span {
            return Err(Status::UnexpectedDimensions);
        }
        let span = match (self.count, candidates.len()) {
            (0, _) | (_, 0) => 0,
            (queries, candidates_count) => {
                let planes = (queries - 1)
                    .checked_mul(scores_query_stride)
                    .ok_or(Status::OverflowRisk)?;
                let rows = (candidates_count - 1)
                    .checked_mul(scores_candidate_stride)
                    .ok_or(Status::OverflowRisk)?;
                planes
                    .checked_add(rows)
                    .and_then(|offset| offset.checked_add(self.widths_count))
                    .ok_or(Status::OverflowRisk)?
            }
        };
        if scores.len() < span {
            return Err(Status::UnexpectedDimensions);
        }

        let engine = self as *mut Self;
        let output = scores.as_mut_ptr();
        let status = with_sequence(candidates, |sequence| unsafe {
            sz_overlap_scores(
                engine,
                sequence,
                output,
                scores_query_stride,
                scores_candidate_stride,
            )
        });
        match status {
            Status::Success => Ok(()),
            error => Err(error),
        }
    }
}

impl Drop for OverlapEngine {
    fn drop(&mut self) {
        unsafe { sz_overlap_engine_free(self as *mut Self) };
    }
}

File: python/stringzilla/overlap.c
/**
 *  @file python/stringzilla/overlap.c
 *  @author Ash Vardanian
 *  @date January 27, 2024
 *  @brief Window overlap between one prepared batch of queries and many collections of candidates.
 */
#include "stringzilla.h"

/** A forest of query trees built once, probed by a fresh collection of candidates each round. */
typedef struct {
    PyObject ob_base;
    sz_overlap_engine_t engine; //< Owned; freed in `tp_dealloc`, rebuilt by a second `__init__`.
    sz_engine_lock_field_       //< Guards the round scratch a call grows.
} OverlapEngine;

#pragma region Construction

/**
 *  @brief Copies the window widths out of a Python sequence of integers.
 *  @return An array of @p count widths from @c malloc, or @c NULL with a Python exception set.
 */
static sz_size_t *parse_window_widths_(PyObject *widths_obj, sz_size_t *count) {
    PyObject *const widths_tuple = PySequence_Tuple(widths_obj);
    if (!widths_tuple) {
        PyErr_SetString(PyExc_TypeError, "widths must be a sequence of integers");
        return NULL;
    }
    Py_ssize_t const widths_count = PyTuple_GET_SIZE(widths_tuple);
    sz_size_t *const widths = (sz_size_t *)malloc((widths_count ? (sz_size_t)widths_count : 1) * sizeof(sz_size_t));
    if (!widths) {
        Py_DECREF(widths_tuple);
        PyErr_NoMemory();
        return NULL;
    }
    for (Py_ssize_t width_index = 0; width_index < widths_count; ++width_index) {
        PyObject *const item = PyTuple_GET_ITEM(widths_tuple, width_index);
        if (!PyLong_Check(item)) {
            Py_DECREF(widths_tuple);
            free(widths);
            PyErr_SetString(PyExc_TypeError, "widths must be a sequence of integers");
            return NULL;
        }
        widths[width_index] = (sz_size_t)PyLong_AsSize_t(item);
        if (PyErr_Occurred()) {
            Py_DECREF(widths_tuple);
            free(widths);
            return NULL;
        }
    }
    Py_DECREF(widths_tuple);
    *count = (sz_size_t)widths_count;
    return widths;
}

static void OverlapEngine_dealloc(OverlapEngine *self) {
    sz_overlap_engine_free(&self->engine);
    Py_TYPE(self)->tp_free((PyObject *)self);
}

static int OverlapEngine_init(OverlapEngine *self, PyObject *args, PyObject *kwargs) {
    Py_ssize_t const positional_count = args ? PyTuple_GET_SIZE(args) : 0;
    if (positional_count < 1 || positional_count > 2) {
        PyErr_Format(PyExc_TypeError, "OverlapEngine() takes 1 to 2 positional arguments, got %zd", positional_count);
        return -1;
    }
    PyObject *const queries_obj = PyTuple_GET_ITEM(args, 0);
    PyObject *widths_obj = positional_count > 1 ? PyTuple_GET_ITEM(args, 1) : NULL;
    if (kwargs) {
        Py_ssize_t keyword_cursor = 0;
        PyObject *key = NULL, *value = NULL;
        while (PyDict_Next(kwargs, &keyword_cursor, &key, &value)) {
            if (PyUnicode_CompareWithASCIIString(key, "widths") != 0) {
                PyErr_Format(PyExc_TypeError, "OverlapEngine() got an unexpected keyword argument '%U'", key);
                return -1;
            }
            if (widths_obj) {
                PyErr_SetString(PyExc_TypeError, "OverlapEngine() got multiple values for argument 'widths'");
                return -1;
            }
            widths_obj = value;
        }
    }
    if (!widths_obj) {
        PyErr_SetString(PyExc_TypeError, "OverlapEngine() needs the window widths to score at");
        return -1;
    }

    sz_sequence_t queries;
    if (sz_py_export_strings(queries_obj, "queries", &queries) != 0) return -1;
    sz_size_t widths_count = 0;
    sz_size_t *const widths = parse_window_widths_(widths_obj, &widths_count);
    if (!widths) return -1;

    sz_engine_lock_(self);
    sz_overlap_engine_free(&self->engine);
    sz_status_t const status = sz_overlap_engine_init_cpu(&queries, widths, widths_count, STRINGZILLA_NULL,
                                                          &self->engine);
    sz_engine_unlock_(self);
    free(widths);
    if (status != sz_success_k) {
        sz_py_raise_status(status, "OverlapEngine()");
        return -1;
    }
    return 0;
}

static char const doc_OverlapEngine_on_gpu[] =                                                 //
    "on_gpu(queries, widths, stream=0) -> OverlapEngine\n"                                     //
    "\n"                                                                                       //
    "Prepare the same forest on a device, on a stream the caller owns.\n"                      //
    "\n"                                                                                       //
    "Args:\n"                                                                                  //
    "  queries (Strs): Texts whose windows are sorted into the forest.\n"                      //
    "  widths (sequence of int): Window widths in bytes.\n"                                    //
    "  stream (int, optional): A `cudaStream_t` as an integer, or 0 for the default stream.\n" //
    "Returns:\n"                                                                               //
    "  OverlapEngine: A device engine whose rounds enqueue and return; join before reading.\n" //
    "Example:\n"                                                                               //
    "  >>> engine = sz.OverlapEngine.on_gpu(sz.Strs(['abcdef']), [3])  # doctest: +SKIP";

static PyObject *OverlapEngine_on_gpu(PyObject *type_obj, PyObject *const *args, Py_ssize_t positional_args_count,
                                      PyObject *args_names_tuple) {
    if (positional_args_count < 1 || positional_args_count > 3) {
        PyErr_Format(PyExc_TypeError, "on_gpu() takes 1 to 3 positional arguments, got %zd", positional_args_count);
        return NULL;
    }
    PyObject *const queries_obj = args[0];
    PyObject *widths_obj = positional_args_count > 1 ? args[1] : NULL;
    PyObject *stream_obj = positional_args_count > 2 ? args[2] : NULL;
    Py_ssize_t const args_names_count = args_names_tuple ? PyTuple_GET_SIZE(args_names_tuple) : 0;
    for (Py_ssize_t keyword_index = 0; keyword_index < args_names_count; ++keyword_index) {
        PyObject *const key = PyTuple_GET_ITEM(args_names_tuple, keyword_index);
        PyObject *const value = args[positional_args_count + keyword_index];
        if (PyUnicode_CompareWithASCIIString(key, "widths") == 0) {
            if (widths_obj) {
                PyErr_SetString(PyExc_TypeError, "on_gpu() got multiple values for argument 'widths'");
                return NULL;
            }
            widths_obj = value;
        }
        else if (PyUnicode_CompareWithASCIIString(key, "stream") == 0) {
            if (stream_obj) {
                PyErr_SetString(PyExc_TypeError, "on_gpu() got multiple values for argument 'stream'");
                return NULL;
            }
            stream_obj = value;
        }
        else {
            PyErr_Format(PyExc_TypeError, "on_gpu() got an unexpected keyword argument '%U'", key);
            return NULL;
        }
    }
    if (!widths_obj) {
        PyErr_SetString(PyExc_TypeError, "on_gpu() needs the window widths to score at");
        return NULL;
    }

    sz_sequence_t queries;
    void *stream = NULL;
    if (sz_py_export_strings(queries_obj, "queries", &queries) != 0) return NULL;
    if (sz_py_export_stream(stream_obj, &stream) != 0) return NULL;
    sz_size_t widths_count = 0;
    sz_size_t *const widths = parse_window_widths_(widths_obj, &widths_count);
    if (!widths) return NULL;

    PyTypeObject *const type = (PyTypeObject *)type_obj;
    OverlapEngine *const self = (OverlapEngine *)type->tp_alloc(type, 0);
    if (!self) {
        free(widths);
        return NULL;
    }
    sz_status_t const status = sz_overlap_engine_init_gpu(&queries, widths, widths_count, STRINGZILLA_NULL, stream,
                                                          &self->engine);
    free(widths);
    if (status != sz_success_k) {
        Py_DECREF(self);
        sz_py_raise_status(status, "OverlapEngine.on_gpu()");
        return NULL;
    }
    return (PyObject *)self;
}

#pragma endregion Construction

#pragma region Operations

static char const doc_OverlapEngine_scores[] =                                             //
    "scores(candidates, out) -> None\n"                                                    //
    "\n"                                                                                   //
    "Score every prepared query against every candidate, at every width, into `out`.\n"    //
    "\n"                                                                                   //
    "Args:\n"                                                                              //
    "  candidates (Strs): Texts whose windows probe the forest.\n"                         //
    "  out (buffer): Writable 3-D buffer of 32-bit floats, at least\n"                     //
    "    (len(queries), len(candidates), len(widths)), contiguous along its width axis.\n" //
    "Example:\n"                                                                           //
    "  >>> engine = sz.OverlapEngine(sz.Strs(['abcdef']), [3])\n"                          //
    "  >>> out = memoryview(bytearray(4)).cast('f', (1, 1, 1))\n"                          //
    "  >>> engine.scores(sz.Strs(['abcdef']), out)\n"                                      //
    "  >>> out[0, 0, 0]\n"                                                                 //
    "  1.0";

static PyObject *OverlapEngine_scores(OverlapEngine *self, PyObject *const *args, Py_ssize_t positional_args_count,
                                      PyObject *args_names_tuple) {
    if (positional_args_count < 1 || positional_args_count > 2) {
        PyErr_Format(PyExc_TypeError, "scores() takes 1 to 2 positional arguments, got %zd", positional_args_count);
        return NULL;
    }
    PyObject *const candidates_obj = args[0];
    PyObject *out_obj = positional_args_count > 1 ? args[1] : NULL;
    Py_ssize_t const args_names_count = args_names_tuple ? PyTuple_GET_SIZE(args_names_tuple) : 0;
    for (Py_ssize_t keyword_index = 0; keyword_index < args_names_count; ++keyword_index) {
        PyObject *const key = PyTuple_GET_ITEM(args_names_tuple, keyword_index);
        PyObject *const value = args[positional_args_count + keyword_index];
        if (PyUnicode_CompareWithASCIIString(key, "out") != 0) {
            PyErr_Format(PyExc_TypeError, "scores() got an unexpected keyword argument '%U'", key);
            return NULL;
        }
        if (out_obj) {
            PyErr_SetString(PyExc_TypeError, "scores() got multiple values for argument 'out'");
            return NULL;
        }
        out_obj = value;
    }
    if (!out_obj) {
        PyErr_SetString(PyExc_TypeError, "scores() needs an `out` buffer to write into");
        return NULL;
    }
    if (!self->engine.memory) {
        PyErr_SetString(PyExc_ValueError, "OverlapEngine holds no prepared queries");
        return NULL;
    }

    sz_sequence_t candidates;
    if (sz_py_export_strings(candidates_obj, "candidates", &candidates) != 0) return NULL;

    sz_size_t const extents[3] = {self->engine.count, candidates.count, self->engine.widths_count};
    sz_size_t strides[3];
    Py_buffer out_view;
    if (sz_py_export_output_buffer(out_obj, "out", (Py_ssize_t)sizeof(sz_f32_t), 3, extents, &out_view, strides) != 0)
        return NULL;
    if (strides[2] != 1) {
        PyErr_SetString(PyExc_ValueError, "out must be contiguous along its width axis");
        PyBuffer_Release(&out_view);
        return NULL;
    }

    sz_engine_lock_(self);
    sz_status_t const status = sz_overlap_scores(&self->engine, &candidates, (sz_f32_t *)out_view.buf, strides[0],
                                                 strides[1]);
    sz_engine_unlock_(self);
    PyBuffer_Release(&out_view);
    if (status != sz_success_k) {
        sz_py_raise_status(status, "scores()");
        return NULL;
    }
    Py_RETURN_NONE;
}

#pragma endregion Operations

#pragma region Type Registration

static char const doc_OverlapEngine[] =                                                          //
    "OverlapEngine(queries, widths)\n"                                                           //
    "\n"                                                                                         //
    "Hash a batch of queries into one window forest and probe it with many candidates.\n"        //
    "\n"                                                                                         //
    "A window is a fixed-width byte n-gram, and a score is the share of a candidate's windows\n" //
    "that the query also spells, in [0, 1]. The score is asymmetric in the two sides.\n"         //
    "\n"                                                                                         //
    "Args:\n"                                                                                    //
    "  queries (Strs): Texts whose windows are sorted into the forest.\n"                        //
    "  widths (sequence of int): Window widths in bytes, the last axis of every output.\n"       //
    "Example:\n"                                                                                 //
    "  >>> engine = sz.OverlapEngine(sz.Strs(['abcdef']), [3, 4])\n"                             //
    "  >>> out = memoryview(bytearray(8)).cast('f', (1, 1, 2))\n"                                //
    "  >>> engine.scores(sz.Strs(['abcdefg']), out)\n"                                           //
    "  >>> 0.0 <= out[0, 0, 0] <= 1.0\n"                                                         //
    "  True";

static PyMethodDef OverlapEngine_methods[] = {
    {"on_gpu", (PyCFunction)OverlapEngine_on_gpu, STRINGZILLA_METHOD_FLAGS | METH_CLASS, doc_OverlapEngine_on_gpu},
    {"scores", (PyCFunction)OverlapEngine_scores, STRINGZILLA_METHOD_FLAGS, doc_OverlapEngine_scores},
    {NULL, NULL, 0, NULL},
};

PyTypeObject OverlapEngineType = {
    PyVarObject_HEAD_INIT(NULL, 0).tp_name = "stringzilla.OverlapEngine",
    .tp_doc = doc_OverlapEngine,
    .tp_basicsize = sizeof(OverlapEngine),
    .tp_itemsize = 0,
    .tp_flags = Py_TPFLAGS_DEFAULT,
    .tp_new = PyType_GenericNew,
    .tp_init = (initproc)OverlapEngine_init,
    .tp_dealloc = (destructor)OverlapEngine_dealloc,
    .tp_methods = OverlapEngine_methods,
};

#pragma endregion Type Registration
